### ISSUE TITLE
Get rid of the "mode" key ...

### ISSUE THREAD

**Issue by pvretano** [opengeospatial/ogcapi-processes#127]
Right now the mode key is used to trigger sync or async execution.  I think we should get rid of the mode key altogether and instead do the following:

- The default execution mode should be what we now call "auto".
- The client submits a request and the server can response in one of two ways
- 1) The server can respond with a 200 and the body contains a response document.
- 2) If the server determines that process execution will take a long time, the server can switch to async processing and respond with a 202.  In this case, the body contains a status message that contains a rel="monitor" link to allow the client to monitor execution by polling.  The body can optionally contain a rel="cancel" link to allow the client to cancel the job.

- The client can also explicitly request asynchronous execution by specifying a "responseHandler" parameter on the request.
- The details of async processing can be found here: https://github.com/opengeospatial/ogcapi-common/tree/master/proposals/simple-async

In both cases the /job endpoint is used to manage each execution job.  Of course, for pure synchronous execution the specification can recommend that the server create an implicit job and let the results persist for a bit but that would a recommendation, not a requirement.

See issue #124 for additional discussion.

NOTE: I am not proposing this change for the pending version of the specification put perhaps a subsequent update.  As has been mentioned in several issues (including issue #124) this change is the more RESTful way to handle sync and async process execution.

**Comment by jerstlouis** [comment]
What if `mode` is made optional and defaults to auto?

I believe it's important that a client be able to still explicitly request synchronous processing because that is all it is able to handle, and the server can return an error if it cannot process that request synchronously.

I also think it's important that `/jobs` and `/jobs/{jobId}` end-point not be required if the server only implements synchronous processing and no job listing / job control conformance classes. (With the exception of keeping `/jobs` as a location to submit the execute request, in case the suggestion to change that to `/processes/{processId}` in #124 is not accepted).

**Comment by pvretano** [comment]
@jerstlouis right now the specification does not make that distinction.  In other words it does not segregate execute between synchronous and asynchronous conformance classes.  That would be a BIG change, I think, and to be honest, MOST (if not all) of the process execution that I have seen in the last few years (especially as related to EO) has be asynchronous.
Even making mode optional and defaulting to auto at this point may be a big change but I will look and see what the impact would be.  We can discuss at the next SWG meeting.

**Comment by jerstlouis** [comment]
@pvretano Right now it does segregate the following conformance classes:

- Job list
- Callback
- Dismiss

And it also allows individual processes to declare whether they support `"async-execute"` and/or `"sync-execute"` in the `"jobControlOptions"`.

So even though there is no synchronous or asynchronous conformance classes, it effectively allow servers to support either or both execution modes.

Regarding EO processing, my view is that we move towards client-driven and targeted processing rather than batch processing, synchronous execution actually becomes much more practical, with clients requesting small areas or tiles to be processed.

Tiles has the extra advantage of being convenient for caching as exact requests will be happening more, plus they are easy to use with parallel/distributed processing.

**Comment by pvretano** [comment]
See: https://github.com/opengeospatial/ogcapi-routes/issues/15

**Comment by gfenoy** [comment]
See also: #30

**Comment by jerstlouis** [comment]
See proposal at https://github.com/opengeospatial/ogcapi-processes/issues/171#issuecomment-824739111

**Comment by fmigneault** [comment]
The first option `The server can respond with a 200 and the body contains a response document.` feels prone to confusion. If 200 is returned while a job and/or response document is created (or accepted for later execution), it is simply the wrong code. But even then, that means the user would always need to make 2 requests to execute each process ? One to fetch the response document and another to actually submit it with sync/async? If so, it is a very confusing way to do it and requires many unnecessary back-and-forth requests. 

The second condition `If the server determines that process execution will take a long time` also doesn't feel reasonable. It is something that more often then not cannot be guessed in advance. Example, image some process that fetches a file and do processing on it. If submitted with a file with tiny area or a massive one, that could make a difference of few seconds to hours if not days. The process cannot know in advance until it fetched the data, so it cannot take the decision on how long it could take. 

I agree that `auto` could be a default of `mode` if not provided, since the results would be the same (the server decides whichever it prefers), but I don't like the idea of removing the possibility from the user to specifically indicate if a process should be `sync` or `async` as needed. The `mode` should definitely remain available. 

IMO `responseHandler` is a bad idea. It seems to come from features that employs GET requests and use of query parameters. Features use case might be valid since it is to retrieve (aka to GET files), but for WPS processes, we are submitting a job. The POST request is already used to (create 201 or accept 202) the submitted job, so it makes more sense for job mode to be provided in the same body. Another bad thing about `responseHandler` is the actual name, that it is not at all intuitive that it controls sync/async execution mode. The execution method doesn't have anything to do with how the *response should be handled*. 

The `cancel` and `monitor` links are good ideas. It helps finding items wherever they might be requested, without the user having to manually generate the URL from a response job UUID for example.

**Comment by jerstlouis** [comment]
@fmigneault Please see proposal in https://github.com/opengeospatial/ogcapi-processes/issues/171#issuecomment-824739111 which discusses a solution using the `Prefer` header approach that we have been discussing more recently, while maintaining the users's capability to select sync or async.

> The process cannot know in advance until it fetched the data, so it cannot take the decision on how long it could take.

These are some of the hurdles that I see the _Workflows_ extension as solving.
A service (including one part of a hop) can request the collection description document and get information like _DomainSet_ and _RangeType_ that can give a pretty good idea of the size of the data. All of this can be done once upon workflow registration / validation, so that e.g. a particular workflow can be associated with a specific bandwidth/processing resource cost, useful for this purpose, but also for estimates/billing, and for planning resource allocation etc. The whole chain validation process should not involve any time-consuming operation, so that the whole workflow can be validated quickly.

**Comment by fmigneault** [comment]
@jerstlouis 
Although Workflows extension could potentially resolve the problem (I'm not sure as I am not familiar with it for nested processes), it remains an *optional* extension, so we cannot rely on it to resolve other problems from the core WPS. 

In CRIM implementation, we employ a complete different approach to handle workflows, so I don't see that extension being implemented on our side anytime soon. Submitted inputs can also vary a lot from a request to another, so even monitoring "usual" bandwidth is not perfectly reliable. Since the only thing submitted is a link to some file that gets dispatched to a remote worker, I cannot guarantee how long it will take to download and process, not to mention if further delays occur if the worker is last in a processing queue.

Thanks for the notification about the other discussion. I added my feedback.

**Comment by jerstlouis** [comment]
@fmigneault It should still be possible for you to implement the server-side decision by waiting for the `Prefer`red timeout before returning a 202 rather than a 200 though, right?

Regarding employing a different approach to _Workflows_, the _Workflows_ extension is really complementary to your current approach using CWL, in that it allows workflows to be put together on the client side. The processes referred to by those client-driven workflows can themselves be workflows defined by CWL on the server side.

**Comment by fmigneault** [comment]
@jerstlouis 
So far, all our processes seem to take "long enough" to always default to 202 and async execution. This is why I have a concern with any SHALL or MUST in the specification that enforces some kind of `sync` implementation. Our current implementation defaults `auto` to `async`, and explicitly return not implemented when `sync` is requested. 

I think OGC should remain neutral about `sync` vs `async` as long as one is implemented. Supporting both is better, but shouldn't be mandatory.

**Comment by jerstlouis** [comment]
@fmigneault I completely agree.

As I clarified that SHALL is not intended to preclude the possibility of the server returning "not implemented / supported". I'm not sure what the proper HTTP response is, but I believe `501 - Not Implemented` is specifically when not supporting a given HTTP _method_, and I could not find any 4xx code specific to that situation, so I suggested a general 400.

**Comment by fmigneault** [comment]
@jerstlouis  
I believe `auto` could be resolved as either `sync` or `async`, and this is why the proposal should not enforce any requirement to require `Prefer` nor reject requests if it was omitted when I desire `async` to always be my default. 
`Prefer` should be added only if you desire the opposite behaviour the implementation automatically chooses, or if you want to make sure that the picked mode matches your preference if both are supported. 
If `auto` was forced to `sync` by default when omitted, as the SHALL statement proposes, then there is no need for `auto` at all (default would simply be `sync` directly), and the `auto` looses its meaning of selecting automatically the preferred method.

**Comment by jerstlouis** [comment]
@fmigneault the proposal for `auto` came before consideration of the `Prefer` header.
The header allows the client to say `Prefer: wait=10`, and then the server can automatically decide to respond asynchronously or synchronously based on whether it thinks (or realizes) that the request will take more than 10 seconds to process.

**Comment by cristianoManuelLopes** [comment]
[Same comment as on #171 ](https://github.com/opengeospatial/ogcapi-processes/issues/171#issuecomment-825482806)

**Comment by fmigneault** [comment]
@jerstlouis 
I really like the `Prefer` header proposal in itself. 
I simply don't want the default behaviour of `auto` or `Perfer` (whichever is selected or both) to be `sync` when it is not provided.

**Comment by jerstlouis** [comment]
@fmigneault The thing is this `Prefer` would be across all OGC APIs.
Maybe it could be left up to the specifications (e.g. _Processes_ or _Features_) to say what the default behavior should be when the `Prefer` header is not provided...

**Comment by fmigneault** [comment]
According to the referenced [RFC7240](https://tools.ietf.org/html/rfc7240#section-2) for `Prefer`, the expected behaviour is that it should be the implementation that takes the decision to respect it or not. If that header and RFC is used, I believe all OGC-APIs should abide to that definition. Custom behaviour going against the RFC will cause even more confusion regarding what to expect as a result. 

> 2.  The Prefer Request Header Field
   The Prefer request header field is used to indicate that particular server behaviors are preferred by the client but are not required for successful completion of the request.

**Comment by pvretano** [comment]
At the last SWG meeting (03-MAY-2021) we decided to reject the PR associated with this issue that would get rid of the mode key and instead rely on standard HTTP mechanisms to trigger asynchronous execution.  I think that was probably the correct decision at this point.  

However, in order to bring OGC API PROCESSES into alignment with OGC API Routes and other possible processing APIs, perhaps we can consider the middle road which is to make **mode optional** with a **default of auto**.  Actually, I would say that the default should be sync since outside of OGC that is the most common mode if client/server interaction but I understand that in our domain the most common execution mode is async.

**Comment by fmigneault** [comment]
Since there will be always disagreements depending on use cases, *default of auto* seems the right way to go. A mention that it is an implementation choice seems adequate since that's what actually causes the different viewpoints.

**Comment by jerstlouis** [comment]
Update: at this morning SWG meeting we agreed to merge PR #82 that makes mode optional, and the default the first option listed in the process description _jobControlOptions_ list.

I believe this addresses most of the concerns raised in this issue. It provides a deterministic way for a client to force the server to execute synchronously or asynchronously.

I also think we could easily re-introduce support for the **_Prefer_** header (and therefore a client preference still considering that the _server knows best_), this time in a completely non-controversial way, by amending @pvretano 's new requirement from the PR to say:

> The first execution mode specified in the array of job control options in the process description SHALL be the default execution mode, unless the request includes a **_Prefer_** header indicating a preference, in which case the default SHALL be `auto`. In that case, the server SHOULD consider the client preference in determining the execution mode.

**Comment by pvretano** [comment]
@jerstlouis since we don't discuss the **Prefer** header at all we should not introduce it in this limited way now.  Rather we should decided to get rid of mode and rely on HTTP methods completely; since we already decided not to do that now, the Prefer header will have to wait.

**Comment by jerstlouis** [comment]
@pvretano I thought this could be completely non-controversial, but since you're giving me a thumb down let's have that discussion again after 1.0 :)

**Comment by pvretano** [comment]
@jerstlouis after 1.0 I would prefer to reintroduce the issue of getting rid of the `mode` flag altogether and instead rely on `Prefer` header completely (i.e. the default execution mode is `sync` unless the `Prefer` header is specified in which case it would be `auto`).
If we even mention `Prefer` now then we would need to completely describe the behavior and include abstract tests.  For example, what happens if the `mode` flag is set of `sync` but the `Prefer` header is also specified and it is set to `respond-async`?  The answer it that the `mode` flag would take precedence but my point is that we still need to consider all those cases AND introduce requirements to deal with the AND introduce abstract tests to validate the requirements.

**Comment by jerstlouis** [comment]
@pvretano (after 1.0)

> For example, what happens if the mode flag is set of sync but the Prefer header is also specified and it is set to respond-async?   The answer it that the mode flag would take precedence

I believe the existing requirement covered that, since the Prefer header is only mentioned in that requirement about the "default" mode. So that would imply that the Prefer header is otherwise ignored.

In terms of Abstract Tests, I thought we could get away with it since it is impossible to test "auto" or "prefer" since the server do whatever it pleases anyways.
But you may be right that just mentioning Prefer header introduces ambiguity that should be covered by additional tests...

> I would prefer to reintroduce the issue of getting rid of the mode flag altogether and instead rely on Prefer header completely (i.e. the default execution mode is sync unless the Prefer header is specified in which case it would be auto).

That would make sense to me if we could agree to that, but agreeing to this after 1.0 seems like it would be a breaking change: existing client that specifies "mode" : "async" without a Prefer header can currently rely on getting a 201 back.

**Comment by pvretano** [comment]
@jerstlouis , that was not an invitation to reopen the debate about this issue.  My point was that if we introduce the `Prefer` header in any way then we will need to fully specify its behavior.  This is something that the SWG has already rejected for now.

**Comment by jerstlouis** [comment]
@pvretano Sorry, I didn't mean to reopen the debate, I know. :) We were discussing the post 1.0 considerations...

Breaking changes are the worst and mean a major version bump, so that's just something to keep in mind. I don't think getting a 1.0 out and then coming up with a non-interoperable 2.0 version soon afterwards is what members insisting on getting this specification out ASAP really want.

That's why I brought this up, because I don't think the approach you suggest will be feasible after 1.0, since it's likely a breaking change.

**Comment by pvretano** [comment]
@jerstlouis ... sigh!  As I said, you proposed the change, I wrote the PR and we tried to convince the SWG to adopt the use of Prefer.  There was no consensus and the proposal was rejected.  If you want to reintroduce the issue in the SWG, go ahead and I will support you (of course since I wrote the PR).  But for now, I am going to concentrate on the stuff we have agreed on so that the document is ready for the upcoming TC.
